File: app/src/main/java/com/example/coinstest/framework/SoapCurrencyInRubDataSource.java
package com.example.coinstest.framework;

import com.example.coinstest.data.CurrencyInRubDataSource;
import com.example.coinstest.domain.CurrencyInRub;
import com.example.coinstest.framework.soap.SoapCbrWebService;

import java.time.LocalDate;


public class SoapCurrencyInRubDataSource implements CurrencyInRubDataSource {

    private SoapCbrWebService webService = new SoapCbrWebService();

    @Override
    public CurrencyInRub getCurrencyOnDate(LocalDate date) {
        return webService.getCurrency("USD", date);
    }

    @Override
    public LocalDate getLastDate() {
        return webService.getLastServerDate();
    }
}

File: app/src/main/java/com/andysklyarov/finnotifyfree/framework/SoapCurrencyInRubDataSource.kt
package com.andysklyarov.finnotifyfree.framework

import com.andysklyarov.finnotifyfree.data.CurrencyInRubDataSource
import com.andysklyarov.finnotifyfree.domain.CurrencyInRub
import com.andysklyarov.finnotifyfree.framework.soap.SoapCbrUtils
import io.reactivex.Single
import java.time.LocalDate

class SoapCurrencyInRubDataSource : CurrencyInRubDataSource {
    override fun getCurrencyOnDate(currencyCode: String, date: LocalDate): Single<CurrencyInRub> {
        return SoapCbrUtils.getCurrencyOnDate(currencyCode, date)
    }

    override fun getCurrencyOnPreviousDate(
        currencyCode: String,
        date: LocalDate
    ): Single<CurrencyInRub> {
        return SoapCbrUtils.getCurrencyOnPreviousDate(currencyCode, date)
    }

    override fun getLastDate(): Single<LocalDate> {
        return SoapCbrUtils.getLastServerDate()
    }
}